src/encoding/text.rs

```diff
@@ -375,11 +375,15 @@ impl<'a> MetricEncoder<'a> {
         }
 
         if let Some(labels) = &self.family_labels {
-            if !self.const_labels.is_empty() || additional_labels.is_some() {
-                self.writer.write_str(",")?;
+            let mut string_writer = String::new();
+            labels.encode(LabelSetEncoder::new(&mut string_writer).into())?;
+
+            if !string_writer.is_empty() {
+                if !self.const_labels.is_empty() || additional_labels.is_some() {
+                    self.writer.write_str(",")?;
+                }
+                self.writer.write_str(string_writer.as_str())?;
             }
-
-            labels.encode(LabelSetEncoder::new(self.writer).into())?;
```

Comment:
Sorry for only raising this now.

This change is not ideal, as it doubles the allocations required for `self.family_labels`, i.e. adds unnecessary allocations to the happy path.

I see multiple ways to still remove the bug (i.e. an invalid dangling `,`) without the allocation:

1. Fail to derive `EncodeLabelSet` on an empty `struct`. This is a partial fix only as one could still go through e.g. a `Vec<(String, String)>`.
2. Wrap `self.writer` with a thin wrapper, that simply slips in a `,` in case any of its `std::fmt::Write` methods are called. In other words that adds a `,` at very first in case `family_labels` actually attempts to write anything.

Does the above make sense? I am in favor of (2). What do you think?

Reply:
Yes, that makes perfect sense to me - thanks for the suggestions. Apologies for my inactivity on this PR, but I'll go for option 2 once I get some time to work on this again. The approach seems pretty straightforward to me.